### ISSUE TITLE
COG + Style Expression: Only apply to non no-data values

### ISSUE THREAD

**Issue by m-mohr** [openlayers/openlayers#13588]
I'm trying to make https://openlayers.org/en/latest/examples/cog-stretch.html work for a COG that has no-data values. In the example the alpha channel always gets set to 1, which then doesn't hide any no-data values anymore. It would be helpful if it would be possible to apply a style only to non no-data pixels, I think.

**Comment by mike-000** [comment]
There is an additional band for alpha which makes nodata transparent https://codesandbox.io/s/cog-stretch-forked-krj47y?file=/main.js

**Comment by m-mohr** [comment]
@mike-000 Thank you, I'll try that again. I had this variant done (without normalization), but it did not work for me. Need to check which detail I missed.

**Comment by m-mohr** [comment]
I'm reopening this because I feel like the documentation needs some improvements.

The style documentation says values are between 0...1 but doesn't mention that if normalize is set to false it is not 0...1 but the original range. Also, it was not clear to me that the alpha band in this case can't simply be passed through. You need to divide by 255 (I guess, or is it the max value of the data type?).

**Comment by m-mohr** [comment]
@mike-000 
Yes, but for me it's not obvious from the documentation what I get for ['band', x] when I'm requesting the band value for 
a) alpha, normalized (0..1?)
b) alpha, not-normalized (0..255?)
c) data band, normalized (0..1)
d) data band, not-normalized (from docs sounds like 0..1, but gives original values)

Realizing that b seems to be 0.255 instead of 0..1 took me two working days... :-|

**Comment by tschaub** [comment]
@m-mohr - I agree there is lots of room for improvement - at least in the docs, but also probably in the behavior (though some of that may need to wait for a major release).

Here is a dump of info.  We'll need to do some work to decide how to craft things for the docs and what we might want to change in terms of behavior.

Currently the renderer works with textures in WebGL that are either single byte (`gl.UNSIGNED_BYTE`) or float (`gl.FLOAT`).  In a layer's `style`, the `band` expressions get values directly from the textures.  For byte textures, WebGL does 0..1 normalization.  So `band` expressions for byte textures get 0..1.  For float textures, WebGL doesn't do any normalization.  So `band` expressions for float textures get the "raw" values.

So understanding how a `band` expression is going to behave requires knowing whether the textures are byte or float.  I originally had thought we might normalize all to 0..1, but we could also go the other direction and have the `band` expressions convert to 0..255 for byte data.  This would be a breaking change but might be less surprising.

If you are using a data tile source directly, this is probably more obvious.  When your loader returns `Uint8Array|Uint8ClampedArray|DataView`, you get a `gl.UNSIGNED_BYTE` texture.  When your loader returns `Float32Array`, you get a `gl.FLOAT` texture.

When you are using the GeoTIFF source, if you set `normalize: false` you always get a float texture (even if the underlying data type is single byte).  If don't set `normalize: false`, you always get a single byte texture (regardless of the input data type).  So it should be true that `normalize: false` means the `band` expressions will receive the "raw" data values (see below for detail on the added "alpha" channel) and otherwise they will get 0..1.

In the GeoTIFF source, if you set a `nodata` value or if one is found in the GeoTIFF metadata, an additional band is added to the data tiles.  In cases where all other bands equal the nodata value, the additional data tile band value will be set to `255` (always, regardless of the data type of the source).

So, I think it should be true that if you set `normalize: false` and you have nodata values (either explicitly configured or found in the metadata), a `band` expression that accesses this additional "alpha" band will get values from 0..255.

Another tricky part is the appropriate value ranges for colors.  If your color uses an `array` expression, the returned values should be 0..1 for R, G, B, and the optional A channel.  If your color uses a `color` expression, the returned values should be 0..255 for R, G, and B; and should be 0..1 for the A channel.  This is subtle, but was done to align with CSS colors.

One thing that trips me up when I forget about the difference between using a `color` expression or an `array` expression for the style color is that the alpha value is multiplied by the R, G, and B values to create the final color.  So, if you forget that A for `color` only ranges from 0..1 and accidentally supply a value like 255, the final R, G, and B values will be much higher than you think.  We should likely clamp the A value before multiplying it to avoid this confusion.

Adding the "alpha" band was meant to make things render as you'd expect without any additional work.  I do see how it adds surprises when you don't want the data normalized or if you have custom style expressions.  I think we could make some changes to make this less surprising.

Currently, only the GeoTIFF source knows about "nodata" values - it doesn't pass along any of this information to the data tile source or the renderer.  We could add additional options to the data tile source to specify that one band is an "alpha" band (we would also have to tell it the max value or rework how that is set).  The renderer could then discard all pixels where the alpha band is equal to 0 and otherwise apply an alpha (knowing the max).  This would be nice unless someone was working with a GeoTIFF and they wanted to style nodata values in some other way (not transparent).

We could also make it possible to disable the addition of the alpha band in the GeoTIFF source.  Though I'm pretty sure things wouldn't work with nodata values like `NaN`.

In case it is useful, here are two equivalent renderings of the same GeoTIFF (depends on #13669).  The first example uses the GeoTIFF metadata to get the nodata value (nan) and the band statistics (min: -158, max: 4232).  The second example avoids normalization, handles the nodata (alpha band 0) case explicitly, and applies a linear stretch between min and max values.

```js
// try to do the right thing with the GeoTIFF metadata
const source = new GeoTIFF({sources: [{url: '/data/elevation-f32.tif'}]});
const layer = new TileLayer({source});

const map = new Map({
  target: 'map',
  layers: [layer],
  view: source.getView(),
});
```

```js
// don't do any normalization
const source = new GeoTIFF({
  normalize: false,
  sources: [{url: '/data/elevation-f32.tif'}],
});

const layer = new TileLayer({
  source: source,
  style: {
    color: [
      'case',
      ['==', ['band', 2], 0],
      '#00000000',
      [
        'interpolate',
        ['linear'],
        ['band', 1],
        -158,
        '#000000ff',
        4232,
        '#ffffffff',
      ],
    ],
  },
});

const map = new Map({
  target: 'map',
  layers: [layer],
  view: source.getView(),
});
```

I think a `case` expression at the top level is the clearest way to handle "nodata" values currently.

**Comment by m-mohr** [comment]
Hi @tschaub,

thanks a lot for all this useful information. Highly appreciated! 

So if I try to fill my list from above, do I understand correctly that the following should be true:
a) alpha, normalized (0..1 as floating-point values)
b) alpha, not-normalized (0..255 as floating-point values)
c) data band, normalized (0..1 as floating-point values)
d) data band, not-normalized (original values as floating-point values)

I think a first thing that could be improved in the docs is to replace "Fetched values are in the 0..1 range." with something like "Fetched values are in the 0..1 range if normalized, and otherwise, in the original range." or so. That would already clarify a lot.

The information that non-normalized values are always floating-point values is good to know and the information about the multiplication of RGBA values in color/array expressions already helped me to rule out a bug (effectively by clamping values additionally). This information should probably also be added to the docs.

Regarding the values of the alpha band, I'm a bit puzzled still. I'm using your fixes for #13669 already locally (the CodeSandbox doesn't do that), but I'm seeing some weird behavior in my color/array expressions in both cases. I'm visualizing a one-band GeoTiff (with `normalize: false`) with no-data values, so effectively have two bands in the GeoTiff source. I'm applying a color  style (but also tried array with changes to reflect 0..1) where I'm computing values for the first band three times equally. And then I'm adding an expression for the alpha band (second band), but I get weird behavior for some expressions:
1. Your `['case', ['==', ['band', 2], 0], 0, 1]` works in a way that it only allows fully visible and invisible, but nothing in-between
2. Mike proposed `['/', ['band', 2], 255]` above, but that weirdly gives me different transparency levels for each data pixel.
3. Similarly, I also tried `['clamp', ['band', 2], 0, 1]` which behaves like 2.

Here's a repro: https://codesandbox.io/s/cog-forked-f5wgp3?file=/main.js
What do I miss in cases 2 and 3 that they get half-transparent or so? I'd expect that 3 is behaving equally to 1 and that 2 just works with different transparency levels. But it does not although the alpha values seem to only be 0 and 255 respectively.

**Comment by aborruso** [comment]
Hi @tschaub 
I tried to apply what you described, but I can't figure it out.
How do I not see those two big black boxes?

Sample code: https://codesandbox.io/s/read-all-nocolormap-m9ep5i?file=/main.js

Thank you very much

![image](https://user-images.githubusercontent.com/30607/231124181-6ba0aa35-a26c-475f-8618-01903f7c5b5f.png)

**Comment by mike-000** [comment]
It looks like both `0` and `NaN` can mean no data.  Although you can only set one value you could test for both using
`["any", ["==", ["band", 1], 0], ["==", ["band", 2], 0]],`
https://codesandbox.io/s/read-all-nocolormap-forked-nyz73h

**Comment by aborruso** [comment]
@mike-000 you have changed

```
["==", ["band", 2], 0],
```

to

```
["any", ["==", ["band", 1], 0], ["==", ["band", 2], 0]],
```

but I do not understand the meaning. The first one, the one that does not work means that all the band 2 must be transparent.

Your conditions works greatly. I read `["==", ["band", 1], 0]` , which I would read as if also all band 1 pixels should be transparent, but it is not. Can you help me understand the meaning of the expression?

Thank you

**Comment by mike-000** [comment]
`["==", ["band", 1], 0]` means the data is 0 but as it is not NaN band 2 will be 255
`["==", ["band", 2], 0]` means the data is NaN.


When band 2 is 0 band 1 is also set to 0 so you could simply test `["==", ["band", 1], 0]`

But if you had two non-zero no data values you would need two tests.
